### PR TITLE
felix/bpf: CTLB workaround for host and nodeports

### PR DESCRIPTION
**Pull request by tomastigera**
## Description   
    Packets from the host are intercepted at the first device they hit. If
    it is towards the same host, it is the loopback device, otherwise it is
    the main device on the cluster network. At that point we do the DNAT -
    select the backed pod for the nodeport. We assume that there is a
    default route so it hits a device. Otherwise we cannot do much.
    
    We either use FIB to redirect the packet to an outgoing device (if on lo
    or if it should go to a different one than the default) or we redirect
    the packet to the bpfin.cali iface to turn the packet around into the
    host address namespace that can route the packet. If an overlay is used,
    we SNAT the packet so that it has the tunnel's source and we redirect it
    to the tunnel.

## Related issues/PRs

refs https://github.com/projectcalico/calico/issues/4509

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

File: felix/bpf-gpl/fib.h [hunk 1]
```diff
@@ -21,6 +21,17 @@
 
 static CALI_BPF_INLINE bool fib_approve(struct cali_tc_ctx *ctx, __u32 ifindex)
 {
+#ifdef UNITTEST
+	/* Let's assume that unittest is setup so that WEP's are ready - for UT simplicity */
+	return true;
+#else
+	/* If we are turnign packets around on lo to a remote pod, approve the
```
Comment (fasaxc) [inline felix/bpf-gpl/fib.h:28]:
```suggestion
	/* If we are turning packets around on lo to a remote pod, approve the
```

File: felix/bpf-gpl/tc.c [hunk 2]
```diff
@@ -64,11 +64,37 @@ static CALI_BPF_INLINE int calico_tc(struct __sk_buff *skb)
...
 	}
 
+	if (CALI_F_NAT_IF) {
+		switch (skb->mark) {
+		case CALI_SKB_MARK_BYPASS:
+			/* We are turning a packet around to a local WEP using bpfnat
+			 * iface, the WEP should do normal processing.
+			 */
+			{
+				__u32 mark = 0;
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:79]: Why the extra variable?

File: felix/bpf-gpl/tc.c [hunk 3]
```diff
@@ -467,7 +492,24 @@ static CALI_BPF_INLINE void calico_tc_process_ct_lookup(struct cali_tc_ctx *ctx)
 		}
 	}
 
-	if (rt_addr_is_local_host(ctx->state->post_nat_ip_dst)) {
+	if (rt_addr_is_local_host(ctx->state->ip_src)) {
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:495]: Should this be limited to program types where the packet is leaving the host?  Don't want the outbound failsafe to kick in on a (spoofed) inbound packet.

File: felix/bpf-gpl/tc.c [hunk 4]
```diff
@@ -476,16 +518,31 @@ static CALI_BPF_INLINE void calico_tc_process_ct_lookup(struct cali_tc_ctx *ctx)
...
+			if (CALI_F_LO) {
+				CALI_DEBUG("NP redir remote on LO\n");
+				ctx->state->flags |= CALI_ST_CT_NP_LOOP;
+			} else if (CALI_F_MAIN && cali_rt_is_tunneled(dest_rt)) {
+				CALI_DEBUG("NP redir remote on HEP to tunnel\n");
+				ctx->state->flags |= CALI_ST_CT_NP_LOOP;
+			}
+			ctx->state->flags |= CALI_ST_CT_NP_REMOTE;
+			ctx->state->pol_rc = CALI_POL_ALLOW;
+			/* XXX We need to do HEP policy here, where we do NAT. */
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:541]: Bit confused by setting the `pol_rc = CALI_POL_ALLOW` and then this comment.  But then you fall through to `do_policy`

File: felix/bpf-gpl/tc.c [hunk 5]
```diff
@@ -476,16 +518,31 @@ static CALI_BPF_INLINE void calico_tc_process_ct_lookup(struct cali_tc_ctx *ctx)
...
-		if (CALI_F_TO_HEP && is_failsafe_out(ctx->state->ip_proto, ctx->state->post_nat_dport, ctx->state->post_nat_ip_dst)) {
-			CALI_DEBUG("Outbound failsafe port: %d. Skip policy.\n", ctx->state->post_nat_dport);
-			COUNTER_INC(ctx, CALI_REASON_ACCEPTED_BY_FAILSAFE);
+
+	if (CALI_F_TO_HEP && ctx->nat_dest && !skb_seen(ctx->skb) && !(ctx->state->flags & CALI_ST_HOST_PSNAT)) {
+		CALI_DEBUG("Host accesses nodeport backend %x:%d state->flags 0x%x\n",
+			   bpf_htonl(ctx->state->post_nat_ip_dst), ctx->state->post_nat_dport,
+			   ctx->state->flags);
+		if (cali_rt_flags_local_workload(dest_rt->flags)) {
+			CALI_DEBUG("NP redir on HEP - skip policy\n");
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:527]: Is this the first program to see this traffic?  I think "wildcard" HEP egress policy should apply from host to local workload even if going via a nodeport.

File: felix/bpf-gpl/tc.c [hunk 6]
```diff
@@ -1226,6 +1292,33 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct cali_tc_ctx *ctx
 		fib = false; /* Disable FIB because we want to drop to iptables */
 	}
 
+	if (CALI_F_TO_HEP && !skb_seen(skb)) {
+		CALI_DEBUG("Host accesses nodeport backend %x:%d\n",
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:1296]: Not sure I follow; looks like there are several codepaths that lead here (via `goto allow`) and not all are nodeports?

File: felix/fv/bpf_test.go [hunk 7]
```diff
@@ -2493,43 +2494,116 @@ func describeBPFTests(opts ...bpfTestOpt) bool {
...
+										},
+									},
 								}
 								w00Slector := fmt.Sprintf("name=='%s'", w[0][0].Name)
 								pol.Spec.Selector = w00Slector
 
 								pol = createPolicy(pol)
 							})
 
-							if testOpts.connTimeEnabled {
```
Comment (fasaxc) [inline felix/fv/bpf_test.go:2503]: :tada:

**Review by fasaxc — CHANGES_REQUESTED** [review]
Few initial questions, made me wonder if I was missing something so thought it best to discuss before I do more digging

File: felix/bpf/libbpf/libbpf.go [hunk 8]
```diff
@@ -307,6 +307,7 @@ func TcSetGlobals(
 	hostTunnelIP uint32,
 	flags uint32,
 	wgPort uint16,
+	natin, natout uint32,
```
[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/bpf/libbpf/libbpf.go:310]: yeah, it is annoying to add the args :+1:

File: felix/bpf-gpl/tc.c (hunk 3 above)
Comment (fasaxc) [reply felix/bpf-gpl/tc.c:492]: Ah OK, there's a CALI_F_TO_HEP condition inside this but it doesn't cover the `ctx->state->flags |= CALI_ST_SRC_IS_HOST;` line.

File: felix/bpf-gpl/tc.c (hunk 5 above)
Comment (fasaxc) [reply felix/bpf-gpl/tc.c:524]: I was wrong, wildcard HEPs don't affect host to own pod (and we have FVs that assert that)

File: felix/bpf-gpl/tc.c [hunk 9]
```diff
@@ -1226,6 +1291,39 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct cali_tc_ctx *ctx
...
+					state->ct_result.ifindex_fwd = NATIN_IFACE  ;
+					CALI_DEBUG("NP remote WL %x:%d on LO or main HEP\n",
+						bpf_htonl(state->post_nat_ip_dst), state->post_nat_dport);
+					ctx->state->flags |= CALI_ST_CT_NP_LOOP;
+				}
+				ctx->state->flags |= CALI_ST_CT_NP_REMOTE;
+				fib = true; /* Enforce FIB since we want to redirect */
+			}
+		} else {
+			/* Did not find a route for a service BE? Should be just
```
Comment (fasaxc) [inline felix/bpf-gpl/tc.c:1315]: What about out-of-cluster service backends such as the k8s API server?  I don't think they'll have routes

**Review by fasaxc — CHANGES_REQUESTED** [review]
Spotted a couple of points on second read through.  I'm still wary of accept_local; could we get rid of that if we assume bpf_redirect_neigh?

**Comment by tomastigera** [comment]
> Spotted a couple of points on second read through. I'm still wary of accept_local; could we get rid of that if we assume bpf_redirect_neigh?

Not sure as if a pod is responding to local host it needs to revers DNAT and the resulting packet may have the local host's source if it was accessing the pod via local nodeport :sweat_smile:  We may be able to deliver straight to a socket but I did not explore that concept yet.

**Review by fasaxc — APPROVED** [review]

**Pull request merged**